Final/TCPServer.cpp

```diff
@@ -0,0 +1,122 @@
+        error("ERROR listener failed");
+
+    while (true)
+    {
+        // Reset read_fds
+        fd_set read_fds;
+        FD_ZERO(&read_fds);
+        FD_SET(sock_fd, &read_fds);
+        int maxfd = 0;
+        std::function<bool (node *)> iter_full = [&read_fds, &maxfd] (node *cur) {
+            int data = cur->data;
+            FD_SET(data, &read_fds); // populate read_fds from open_conns
+            if (data > maxfd) // calculate maxfd
+                maxfd = data;
+            return true;
+        };
+        open_conns->for_each(iter_full); // iterate through entire list
+
+        if (sock_fd > maxfd)
+            maxfd = sock_fd;
+
+        std::cout << "Attempting to Select . . ." << std::endl;
+        if (select(maxfd + 1, &read_fds, NULL, NULL, NULL) < 0) // sleeps until there is an available connection and then determines if it must be accepted or read from
+            error("My error message");
+        std::cout << "Connection Selected" << std::endl;
+
+        if (FD_ISSET(sock_fd, &read_fds))
+            accept_connection();
+        else
+            handle_connection(read_fds, handler);
+    }
+}
+
+//Pre: sock_fd is the socket bound to this server
+//Post: accept connection and add file descriptor of connected client to open_conns
+void TCPServer::accept_connection()
+{
+    // Step 4: Accecpt
+    struct sockaddr_in remote_host; // a pointer to a sockaddr struct that will be filled in with the address information of the remote host
+    socklen_t remote_len = sizeof(remote_host); // the size of the sockaddr struct in bytes
+    std::cout << "Attempting to Accept . . ." << std::endl;
+    int remotefd = accept(sock_fd, (struct sockaddr *) &remote_host, &remote_len); // dequeue from sock_fd and accept connection through remotefd
+    if (remotefd < 0)
+        error("ERROR accepting request");
+    std::cout << "Connection Accecpted" << std::endl;
+    open_conns->push(remotefd);
+}
+
+//Pre: open_conns contains fd that is also in with read_fds
+//Post: determine which fd has been updated and echo data
+void TCPServer::handle_connection(fd_set read_fds, std::function<bool (int)> handler)
+{
+    int fd;
+    std::function<bool (node *)> iter_search = [read_fds, &fd] (node *cur){
+        if (FD_ISSET(cur->data, &read_fds)) { // check if file descriptor is ready to be read from
+            fd = cur->data;
+            return false;
+        }
+        return true;
+    };
+    open_conns->for_each(iter_search);  // iterate through fdlist
+
+    if(!fd)
+        error("Connection not in list");
+
+    std::cout << "Handeling Received Data" << std::endl;
+
+    open_conns->remove(fd); //avoid race conditon of fd being checked against twice if handler is a slow enough funtion
+    bool add_back = false;
+    thread_pool->enqueue([fd, handler, &add_back]
+    {
+        if(handler(fd))
+            close(fd);
+        else
+            add_back = true;
+    });
+    if (add_back) 
+        open_conns->push(fd);
```

Comment:
this has to happen inside the thread. remember that this code is running in parallel with the thread, and so there'll be a race condition. if the `handler` takes a while to finish, then `add_back` will always be false when execution gets to here.